Record when a Q was marked as answered

ghassanachi opened the pull request:
Issue: #5 

Add timestamp support for when a question was answered, and sort the questions by answered time on the client.

The proposed changes currently reuse the 'answered' field on questions and switches it from a `bool` to a `u64` timestamp. Since this requires a change of type in the DynamoDb tables, a migration script will need to be applied on existing entries. (This is my first time working with DynamoDb, so not sure how hard this is to do, but the script should be very simple since we can simply set all previously answered questions to the attr time after changing the column type from `bool` to `string`.  

For the payload, similar to how the `who` field is being handled, if the question has not been answered then the field is fully omitted from the response. 

I also made the `/toggle` endpoint return the value that was set on for the property in the following json format (this is to allow the `localAdjustements` property on the client to properly compare the values against the api responses).

```json 
{ "<property>":  "<value>" }
```

PS: Sorting is currently happening client side.

jonhoo commented on client/src/Question.svelte:
```diff
@@ -50,13 +50,13 @@
 		localAdjustments.update(la => {
 			let q = la.remap[question.qid] || {};
-			q[what] = !question[what];
+                        q[what] = res[what];
```
I'm not sure if this will work when `answered` is _unset_, at least if that means that the value is unset in the JSON from the server. It'll mean that `q[what]` gets set to `null` (or `undefined`?), rather than getting removed. That is arguably what we want, but I don't know if the code that _applies_ `localAdjustments` does the right thing if it sees an adjustment that says `["answered"] = null`.

ghassanachi replied:
Yea, this code is relying pretty heavily on Javascript semantics, and takes advantage of the fact that an explicit `undefined` and a missing property will both resolve to undefined (example below): 
```javascript 
let q = {  answered: undefined };
let adj = {}
q.answered === adj.answered // true
```
But the explicit set will still resolve to true in the  following code `"answered" in adj`, so the adjustment code will still run where needed.

In practice I think this still works as expected, but I agree that it's not very clear what is going on. I'll see if I can find a better solution tomorrow.

jonhoo commented on server/src/main.rs:
```diff
@@ -134,7 +144,7 @@ async fn main() -> Result<(), Error> {
             text: String,
             hidden: bool,
-            answered: bool,
+            answered: Option<usize>,
```
This doesn't seem right — `server/src/test.json` is using the original live-ask.com data, which only has a boolean for `answered`.

ghassanachi replied:
I didn't realize where the test data was coming from, and that you wanted to keep it as is.  I'll revert it back to the bool version.

When it comes to the `answered` field column, would you rather use a different column, or just remap to the `answered: true` values coming from `test.json` to a timestamp on the fly.

ghassanachi replied:
I reverted the `test.json` file and updated the code in main to generate the `timestamp` while processing the items in the file (ie: if `answered = true` then add a `answered` timestamp for the question.

jonhoo commented on server/src/main.rs:
```diff
@@ -42,6 +43,15 @@ impl Backend {
     }
 }
 
+fn get_dynamo_timestamp(time: SystemTime) -> AttributeValue {
```
```suggestion
fn to_dynamo_timestamp(time: SystemTime) -> AttributeValue {
```

ghassanachi replied:
Good Suggestion. I updated the signature.

jonhoo commented on server/src/test.json:
```diff
@@ -4,919 +4,839 @@
         "text": "How do you manage your time for work, open source, personal projects, and life?",
         "hidden": false,
-        "answered": true,
+        "answered": 1670131269,
```
I'd like to keep this file as-is, since it's a pure export file from live-ask.com.

ghassanachi replied:
Reverted.

jonhoo commented on server/src/toggle.rs:
```diff
@@ -71,25 +96,40 @@ pub(super) async fn toggle(
+                ToggleRequest::Answered(time) => {
+                    if let Some(time) = time {
+                        Ok(Json(serde_json::json!({
+                            "answered": get_dynamo_timestamp(time)
+                                .as_n().ok()
+                                .and_then(|v| v.parse::<usize>().ok())
```
This feels weird — why go via `AttributeValue` at all here? `time` is already a `SystemTime`, so you can just take `duration_since(UNIX_EPOCH)` and it'll give you the right value.

ghassanachi replied:
Good catch, not sure what I was thinking when writing this. I switched to just using SystemTime for this call.

jonhoo reviewed (changes requested):
Nice, thank you! I left a couple of notes, nothing that should be particular major.

jonhoo commented:
Oh, yeah, and I'd like to see a test that checks that `toggle` actually returns the right value now that it returns a value!

ghassanachi commented:
> Oh, yeah, and I'd like to see a test that checks that `toggle` actually returns the right value now that it returns a value!

I was actually hoping to add this test case, but I wasn't sure the best way to test the return value for the timestamps since that depends on `SystemTime::now()`. Not sure if you have some advice on how to mock the call, or if you have some other idea on how we could restructure the code to not rely on a mocked call (If mocking is a common practice in Rust, maybe a `Crust of Rust` about it could be interesting :p)

ghassanachi commented:
Just finished working on fixing the way we handle `localAdjustments` for the `answered` field. I updated the remap property to the following structure and update the `adjustQuestions` code to handle the condition a little more explicitly.
``` 
qid => {
    hidden: bool,
    answered: {action: "unset"} | {action: "set", value: number}, 
    voted_when: int 
}
```
For now I only used the *patch* syntax for the answered field, since I wanted to keep the changes as minimal as possible, but we could also move the other fields to use the same structure, to keep things a little more aligned. 

I also fixed the server issues you commented on (I left the conversations as unresolved, since I remember you complaining about the way `github` handles that in some of your streams). 

The only thing I haven't tackled yet is adding `tests cases` for the toggle return values. Still not sure how to approach mocking the `SystemTime::now` call, and I'd love some guidance or general direction before diving into it.

jonhoo commented on client/src/List.svelte:
```diff
@@ -153,13 +151,18 @@
+					} else if (patch.action === "set" && !("answered" in q)) {
+						console.info("adjust answered to", patch.value);
+						qs[i].answered = patch.value;
+					}else {
```
```suggestion
					} else {
```

jonhoo commented on client/src/List.svelte:
```diff
@@ -153,13 +151,18 @@
+					if (patch.action === "unset" && "answered" in q) {
+						console.info("remove answered property");
+						delete qs[i].answered;
+					} else if (patch.action === "set" && !("answered" in q)) {
```
This isn't 100% correct, though probably good enough in practice. In theory at least, I could mark as answered, wait for a sync, unmark, then mark again, and local adjustments wouldn't end up being applied to set the new timestamp. I don't think that matters though.

ghassanachi replied:
I initially started to write the code to handle those cases, but it complicated the branches quite a bit for a situation that is unlikely to happen (and if it does happen, it should resolve itself on the following sync since we'll use the timestamp from RawQuestions once localAdjustments is marked as resolved). 

I'll let you decide if want to commit as is. If you would prefer a more robust solution let me know and I can switch it up.

jonhoo replied:
I think as-is is fine!

jonhoo commented on client/src/store.js:
```diff
@@ -4,9 +4,9 @@ const storedVotedFor = JSON.parse(localStorage.getItem("votedFor"));
 votedFor.subscribe(value => {
     if (value) {
-        localStorage.setItem("votedFor", JSON.stringify(value));
+	localStorage.setItem("votedFor", JSON.stringify(value));
```
I'm assuming making this a tab wasn't intentional?

ghassanachi replied:
Not intentional, I think I opened the project in code and it must have done the conversion automatically. I'll revert those  once I get home.

jonhoo reviewed (commented):
Looks great! Just a few more notes.

And yes, thanks for not marking as resolved :)

I agree with you that we may want to move to explicit patch values for other things too later, though let's do that separately.

As for testing, I would test that a) the value is unset/set, and b) if set, that it's within, say, 30s of the expected time. That should be good enough.

ghassanachi commented:
Thanks for the Review! 

Just updated the `toggle` test suite and reverted the indentation changes. 

Since dynamo doesn't have a `Batch Update` functionality, we'll need a little script you can use to migrate the `answered` attribute. 

You're probably more familiar with the DynamoDB api (Given you position and experience :p), but if I have a little time to work on it I'll link a gist to a script you can copy and paste.

jonhoo reviewed (approved):
This looks great now, thanks!

jonhoo commented:
And deployed!
